text/0000-sigstore-integration.md

```diff
@@ -0,0 +1,392 @@
+
+This RFC does not recommend a specific timeframe to enter into Phase 2, but assumes that after some suitable period with the basic signature mechanics in place, Phase 2 can be addressed and TUF implementation can be designed and added.
+
+## Cargo publish flow
+
+![Diagram showing the flow on cargo publish including the authentication, signing and publication to crates.io](https://raw.githubusercontent.com/lulf/rfc-resources/main/sigstore/cargo_sigstore-publish.drawio.png)
+
+The following changes must be made to cargo when publishing a crate:
+
+1. Authenticate the publisher using OIDC to retrieve a token
+1. Generate the ephemeral signing key and certificate request
+1. Generate a certificate signing request and pass with a token to Sigstore Fulcio.
+1. Sign the generated .crate file using the private key
+1. Attach signature, identity and certificate to crates.io publish request
+1. Publish an entry to the Rekor log with the .crate digest and certificate public key
+
+Most of the above is already implemented in https://github.com/sigstore/sigstore-rs.
+
+### crates.io flow
+
+The crates.io sub-flow on publish is described below:
+
+![Diagram showing the flow within crates.io as a crate is accepted and stored](https://raw.githubusercontent.com/lulf/rfc-resources/main/sigstore/cargo_sigstore-crates.io.drawio.png)
+
+The following components need changing for crates.io:
+
+1. Additional columns must be added to the PostgreSQL database
+1. HTTP API must handle a new version of NewCratePublish request with the signature data
+1. The HTTP handler talks to Rekor and verifies that the signature is valid and that owner matches certificate
+
+## Cargo verify flow
+
+The cargo flow during verification is shown below:
+
+![Diagram showing the flow in cargo as crates are downloaded and verified against the signature and certificate](https://raw.githubusercontent.com/lulf/rfc-resources/main/sigstore/cargo_sigstore-verify.drawio.png)
+
+The following changes must be made to cargo when building/verifying a crate:
+
+1. Retrieve the crate contents, signature and certificate.
+1. Verify that the crate contents is signed by the signature.
+1. Verify that the certificate and signature is present in the transparency log.
+
+### Offline flow
+
+The offline verification flow is shown below
+
+![Diagram showing the alternative verification flow using he offline bundles that proves that the entry has been stored in the public log](https://raw.githubusercontent.com/lulf/rfc-resources/main/sigstore/cargo_sigstore-verify-offline.drawio.png)
```

Comment:
this image should not have a transparent background since the black lines are basically invisible on github in dark mode.